app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt

```diff
@@ -0,0 +1,332 @@
+        }
+    }
+
+    @VisibleForTesting
+    private fun dispatchFocusUpdates(
+        focusNode: FocusNode,
+        activeScreen: ActiveScreen,
+        sessionState: SessionRepo.State,
+        pinnedTilesIsEmpty: Boolean,
+        pocketState: PocketVideoRepo.FeedState
+    ): State {
+
+        var newState = _state.value!!
+        when (activeScreen) {
+            ActiveScreen.NAVIGATION_OVERLAY -> {
+                when (focusNode.viewId) {
+                    R.id.navUrlInput ->
+                        newState = updateNavUrlInputFocusTree(
+                                focusNode,
+                                sessionState,
+                                pinnedTilesIsEmpty,
+                                pocketState)
+                    R.id.navButtonReload -> {
+                        newState = updateReloadButtonFocusTree(focusNode, sessionState)
+                    }
+                    R.id.navButtonForward -> {
+                        newState = updateForwardButtonFocusTree(focusNode, sessionState)
+                    }
+                    R.id.pocketVideoMegaTileView -> {
+                        newState = updatePocketMegaTileFocusTree(focusNode, pinnedTilesIsEmpty)
+                    }
+                    R.id.megaTileTryAgainButton -> {
+                        newState = handleLostFocusInOverlay(
+                                focusNode,
+                                sessionState,
+                                pinnedTilesIsEmpty,
+                                pocketState)
+                    }
+                    R.id.home_tile -> {
+                        // If pinnedTiles is empty and current focus is on home_tile, we need to
+                        // restore lost focus (this happens when you remove all tiles in the overlay)
+                        if (pinnedTilesIsEmpty) {
+                            newState = handleLostFocusInOverlay(
+                                    focusNode,
+                                    sessionState,
+                                    pinnedTilesIsEmpty,
+                                    pocketState)
+                        }
+                    }
+                    View.NO_ID -> {
+                        // Focus is lost so default it to navUrlInput and set focused = false
+                        val newFocusNode = FocusNode(R.id.navUrlInput)
+
+                        newState = updateNavUrlInputFocusTree(
+                                newFocusNode,
+                                sessionState,
+                                pinnedTilesIsEmpty,
+                                pocketState,
+                                focused = false)
+                    }
+                }
+            }
+            ActiveScreen.WEB_RENDER -> Unit
+            ActiveScreen.POCKET -> Unit
+            ActiveScreen.SETTINGS -> Unit
+        }
+
+        return newState
+    }
+
+    private fun updateNavUrlInputFocusTree(
```

Comment:
`updateNavUrlInputFocusTree` -> `getNavUrlInputFocusNode`
Or something along those lines.  We aren't mutating any existing object, so update seems like a misnomer to me.

Reply:
Same comment for all the other `update` method names.